### PR TITLE
"Resume" and single-contract deployments via CLI.

### PR DESCRIPTION
**Pull request by KPrasch**
Based on #1130 

Uses Agency and Deployer methods to deploy a single contract by it's registry name form the CLI.  This may act as a good secondary means of managing failed deployments as part of an automated series.

### CODE REVIEW

File: nucypher/blockchain/eth/actors.py [hunk 1]
```diff
@@ -218,7 +218,7 @@ def __get_deployer(self, contract_name: str):
         return Deployer
 
     @staticmethod
-    def __collect_deployment_secret(deployer) -> str:
+    def collect_deployment_secret(deployer) -> str:
```
Comment (tuxxy) [inline nucypher/blockchain/eth/actors.py]: No more privacy?

File: nucypher/blockchain/eth/actors.py [hunk 2]
```diff
@@ -500,8 +503,8 @@ def initialize_stake(self,
         if entire_balance:
             amount = self.token_balance
         if not self.token_balance >= amount:
-            raise self.StakerError(f"Insufficient token balance ({self.token_agent}) "
-                                   f"for new stake initialization of {amount}")
+            raise self.InsufficientTokens(f"Insufficient token balance ({self.token_agent}) "
```
Comment (tuxxy) [inline nucypher/blockchain/eth/actors.py]: :+1:

File: nucypher/blockchain/eth/actors.py (hunk 1 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/actors.py]: Yeah, I needed to use this function outside of the" automated series of deployments" (`deploy_network_contracts`) method below this one in the CLI when deploying an individual upgradeable contract, and collecting it's secret.

File: nucypher/blockchain/eth/clients.py [hunk 3]
```diff
@@ -228,7 +233,7 @@ def check_for_timeout(t):
             last_update = maya.now()
             duration = (last_update - start_time).seconds
             if duration > t:
-                raise self.SyncTimeout
+                return True  # Just continue on.  # TODO: does this make sense?
```
Comment (michwill) [inline nucypher/blockchain/eth/clients.py]: Eh? I don't understand why.
Neither do I understand sleep(0) below

File: nucypher/blockchain/eth/clients.py [hunk 4]
```diff
@@ -394,7 +396,7 @@ def __init__(self,
     def provider_uri(self, scheme: str = None) -> str:
         if not scheme:
             scheme = self.IPC_PROTOCOL
-        if scheme == 'file':
+        if scheme in ('file', 'ipc'):
```
Comment (michwill) [inline nucypher/blockchain/eth/clients.py]: :+1:

File: nucypher/cli/deploy.py [hunk 5]
```diff
@@ -171,42 +170,42 @@ def deploy(action,
...
             else:
-                # Deploy single contract
-                receipts, agent = deployer.deploy_contract(contract_name=contract_name,
-                                                           plaintext_secret=deployment_secret)
+                click.secho(f"Deploying {contract_name}")
+                if contract_deployer._upgradeable:
+                    secret = deployer.collect_deployment_secret(deployer=contract_deployer)
+                    receipts, agent = deployer.deploy_contract(contract_name=contract_name, plaintext_secret=secret)
+                else:
+                    receipts, agent = deployer.deploy_contract(contract_name=contract_name, gas_limit=gas)
```
Comment (michwill) [inline nucypher/cli/deploy.py]: :+1:

File: nucypher/blockchain/eth/clients.py (hunk 3 above)
Comment (KPrasch) [reply nucypher/blockchain/eth/clients.py]: Okay - I restored the exception here.

Comment (KPrasch) [reply nucypher/blockchain/eth/clients.py]: With regard to `time.sleep(0)` - Take a look at this stack overflow: <https://stackoverflow.com/questions/7273474/behavior-of-pythons-time-sleep0-under-linux-does-it-cause-a-context-switch>.  

A comment may be helpful here.

**Comment by michwill** [comment]
:cowboy_hat_face:

File: nucypher/cli/deploy.py [hunk 6]
```diff
@@ -171,42 +170,42 @@ def deploy(action,
...
                 click.secho(message, fg='red', bold=True)
                 raise click.Abort()
             else:
-                # Deploy single contract
-                receipts, agent = deployer.deploy_contract(contract_name=contract_name,
-                                                           plaintext_secret=deployment_secret)
+                click.secho(f"Deploying {contract_name}")
+                if contract_deployer._upgradeable:
+                    secret = deployer.collect_deployment_secret(deployer=contract_deployer)
+                    receipts, agent = deployer.deploy_contract(contract_name=contract_name, plaintext_secret=secret)
```
Comment (cygnusv) [inline nucypher/cli/deploy.py]:
```suggestion
                    receipts, agent = deployer.deploy_contract(contract_name=contract_name, plaintext_secret=secret, gas_limit=gas)
```
It also needs to be passed to `deployer.deploy_network_contracts(secrets=secrets)` in line 243:
https://github.com/nucypher/nucypher/pull/1131/files#diff-4809c009687ec0119c014880f31f9111R243

File: tests/blockchain/eth/clients/test_mocked_clients.py [hunk 7]
```diff
@@ -24,9 +24,10 @@ class MockGanacheProvider:
 
 
 class ChainIdReporter:
+
     # Support older and newer versions of web3 py in-test
     version = 5
-    chainID = 5
+    chainId = hex(5)
```
Comment (cygnusv) [inline tests/blockchain/eth/clients/test_mocked_clients.py]: Just out of curiosity: why hex here?

Comment (KPrasch) [reply tests/blockchain/eth/clients/test_mocked_clients.py]: Just following what geth nodes return as a type on this endpoint for mocking

**Comment by KPrasch** [comment]
- When deploying contracts individually, it's not printing the registry path at the end.
:heavy_check_mark: 

- The gas limit is ignored sometimes: for instance, when deploying all contracts (i.e., no contract name is given), or when a deploying a named contract but it's upgradeable. See my comments above.
Yeah, I am aware of this, I think gas management is generally outside the scope of this PR.  If you prefer a higher degree of precision here I can remove these lines for now.  (#842 )

- Is is possible that nucypher stake list shows stakes for different stakers? I guess that was the point of Stakeholder, but I don't know how to try it.

Yes  - you can run `nucypher stake accounts` or `nucypher stake list`; However, there is not presently an interactive "staker" selection menu.  This is a God suggestion perhaps for an Issue.

**Pull request merged**